src/main/java/org/apache/commons/compress/archivers/zip/ZipFile.java

```diff
@@ -1111,14 +1122,11 @@ public int read() throws IOException {
                 }
                 return -1;
             }
-            synchronized (archive) {
-                archive.position(loc++);
-                int read = read(1);
-                if (read < 0) {
-                    return read;
-                }
-                return buffer.get() & 0xff;
+            int read = read(loc++, 1);
+            if (read < 0) {
```

Comment:
Correct, bodewig. Additionally, note that in these terms the original implementation wasn't thread-safe - in the byte[] read method the loc incrementing is outside the synchronized too.

Anyway, there are two options - either make the full methods synchronized (the single stream would hardly ever be accessed concurrently so keeping smaller scope won't make any difference).

Or, better option would be to keep this implementation unsynchronized as it's wrapped into another decompressing stream anyway and this one already has to be synchronized. The only exception is STORED compression method where we would have to wrap into synchronized proxy.